sdk/test-utils/perfstress/src/testProxyHttpClient.ts

```diff
@@ -0,0 +1,289 @@
+};
+
+/**
+ * Helper class to manage the recording state to make sure the proxy-tool is not flooded with unintended requests.
+ *
+ * => then start record
+ * => run the runAsync
+ * => stop record
+ * => start playback
+ * => stop playback
+ */
+export class RecordingStateManager {
+  public state:
+    | "started-recording"
+    | "stopped-recording"
+    | "started-playback"
+    | "stopped-playback"
+    | undefined;
+
+  /**
+   * validateState
+   */
+  public validateState(
+    currentFlow:
+      | "starting-recording"
+      | "stopping-recording"
+      | "starting-playback"
+      | "stopping-playback"
+  ) {
+    if (currentFlow === "starting-recording") {
+      if (this.state === "started-recording") {
+        throw new Error("Already started recording, should not have called again.");
+      }
+    }
+    if (currentFlow === "stopping-recording") {
+      if (this.state === "stopped-recording") {
+        throw new Error("Already stopped recording, should not have called again.");
+      }
+      if (this.state !== "started-recording") {
+        throw new Error("Please start recording before calling stop.");
+      }
+    }
+    if (currentFlow === "starting-playback") {
+      if (this.state !== "stopped-recording") {
+        throw new Error("Did not stop recording, stop recording before starting playback.");
+      }
+    }
+    if (currentFlow === "stopping-playback") {
+      if (this.state !== "started-playback") {
+        throw new Error("Did not start playback, start playback before calling stop.");
+      }
+    }
+  }
+
+  /**
+   * setState
+   */
+  public setState(
+    state: "started-recording" | "stopped-recording" | "started-playback" | "stopped-playback"
+  ) {
+    this.state = state;
+  }
+}
+
+export class TestProxyHttpClientV1 extends DefaultHttpClient {
+  private _uri: string;
+  private _httpClient: DefaultHttpClient;
+  private _recordingId?: string;
+  public _mode!: string;
+  private stateManager: RecordingStateManager = new RecordingStateManager();
+
+  constructor(uri: string) {
+    super();
+    this._uri = uri;
+    this._httpClient = new DefaultHttpClient();
+  }
+
+  async sendRequest(request: WebResourceLike): Promise<HttpOperationResponse> {
+    if (this._recordingId && (this._mode === "record" || this._mode === "playback")) {
+      request.headers.set("x-recording-id", this._recordingId);
+      request.headers.set("x-recording-mode", this._mode);
+      request.headers.set("x-recording-remove", "false");
+
+      const upstreamUrl = URLBuilder.parse(request.url);
+      const redirectedUrl: URLBuilder = URLBuilder.parse(request.url);
+
+      const providedUrl = URLBuilder.parse(this._uri);
+      redirectedUrl.setHost(providedUrl.getHost());
+      redirectedUrl.setScheme(providedUrl.getScheme());
+      redirectedUrl.setPort(providedUrl.getPort());
+      upstreamUrl.setPath(undefined);
+      request.headers.set("x-recording-upstream-base-uri", upstreamUrl.toString());
```

Comment:
If JS is similar to .NET, you need to ensure this header is only set once, since this codepath may execute multiple times on retries:

https://github.com/Azure/azure-sdk-for-net/pull/22621/files#diff-b2fb8a7a06a2f30913d5933b577bf9e6763764c19af5e841ec450c2b360bfe2aR47

Reply:
Done!